_source/_posts/2021-11-16-k8s-security-best-practices.md

```diff
@@ -0,0 +1,148 @@
+- Prevent loading unwanted kernel modules in the containers. These can be restricted using rules in `/etc/modprobe.d/kubernetes-blacklist.conf` of the node or by uninstalling the unwanted modules from the node.
+- Use [official verified images](https://docs.docker.com/docker-hub/official_images/) for popular software. Use a trusted registry for non-official images and always verify the image publisher
+- Use [Docker Bench for Security](https://github.com/docker/docker-bench-security) to audit your container images
+- Use [Pod security policies](https://kubernetes.io/docs/concepts/policy/pod-security-policy/) to limit a container's access to the host further
+
+You can read more about it in our "[Container Security: A Developer Guide](https://developer.okta.com/blog/2019/07/18/container-security-a-developer-guide)".
+
+## 7. Control traffic between pods and clusters
+
+Generally, pods within the same cluster will be able to communicate with each other, and if you have multiple clusters in the same network, there may be traffic between them as well. Do not let this all open, as it could lead to a compromised cluster when another in the network is affected. Use [Kubernetes network policies](https://kubernetes.io/docs/concepts/services-networking/network-policies/) to control traffic between pods and clusters and allow only necessary traffic.
+
+## 8. Use namespaces to isolate workloads
+
+Do not run all your workloads in a single namespace. Isolating workloads in different namespaces based on business needs will make it more secure and easier to manage with RBAC. This way, you can fine-tune RBAC even further to let users access only what they need to see. You can also use Kubernetes network policies to isolate traffic between namespaces where applicable.
+
+## 9. Limit resource usages
+
+As with securing APIs and the cluster itself, it is also essential to set resource limits on how much CPU, memory, and persistent disk space is used by namespaces and resources. This secures your cluster from denial of service attacks when a particular container uses up all the resources. [Resources quotas](https://kubernetes.io/docs/concepts/policy/resource-quotas/) and [limit ranges](https://kubernetes.io/docs/tasks/administer-cluster/manage-resources/memory-default-namespace/) can be used to set limits at the namespace level, and [Requests and limits](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/) can be used to set resource limits at container level as well.
+
+## 10. Use monitoring tools to monitor all traffic and enable audit logging
+
+Finally, it is also extremely important to monitor and audit your clusters. [Enable audit logging for the cluster](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/) and use monitoring tools to keep an eye on the networking traffic to, from and within a cluster. Monitoring can be done using open-source tools, like Prometheus, Grafana, or with proprietary tools.
```

Comment:
got any suggestions about audit cadence?

Reply:
I'm not too familiar with that so probably not the right person to give advice on that. If you have something in mind, do suggest